File: _old_csharp/addons/terrabrush/Scripts/Tools/HoleTool.cs
#if TOOLS
using System;
using System.Collections.Generic;
using System.Linq;
using Godot;

namespace TerraBrush;

public class HoleTool : ToolBase {
    private HashSet<ZoneResource> _sculptedZones;

    protected override bool ApplyResolution => true;

    public HoleTool(TerraBrush terraBrush) : base(terraBrush) {}

    protected override ImageTexture GetToolCurrentImageTexture(ZoneResource zone) {
        return zone.HeightMapTexture;
    }

    public override void BeginPaint() {
        base.BeginPaint();

        _sculptedZones = new HashSet<ZoneResource>();
    }

    public async override void EndPaint() {
        base.EndPaint();

        _sculptedZones = null;

        _terraBrush.ClearObjects();
        await _terraBrush.CreateObjects();
    }

    public override void Paint(TerrainToolType toolType, Image brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
        ForEachBrushPixel(brushImage, brushSize, imagePosition, (imageZoneInfo, pixelBrushStrength) => {
            var currentPixel = imageZoneInfo.Image.GetPixel(imageZoneInfo.ZoneInfo.ImagePosition.X, imageZoneInfo.ZoneInfo.ImagePosition.Y);
            var holeValue = toolType == TerrainToolType.HoleAdd ? 1f : 0f;

            var newValue = new Color(
                currentPixel.R,
                Math.Max(0.0f, Math.Min(1.0f, Mathf.Lerp(currentPixel.G, holeValue, (float) Math.Ceiling(pixelBrushStrength) * (brushStrength * 100.0f)))),
                currentPixel.B,
                currentPixel.A
            );

            imageZoneInfo.Image.SetPixel(imageZoneInfo.ZoneInfo.ImagePosition.X, imageZoneInfo.ZoneInfo.ImagePosition.Y, newValue);
            _sculptedZones.Add(imageZoneInfo.Zone);
        });

        _terraBrush.TerrainZones.UpdateHeightmaps();
    }
}
#endif

File: src/editor_tools/hole_tool.cpp
#include "hole_tool.h"

using namespace godot;

void HoleTool::_bind_methods() {}

HoleTool::HoleTool() {}

HoleTool::~HoleTool() {}

bool HoleTool::getApplyResolution() const {
    return true;
}

void HoleTool::beginPaint() {
    ToolBase::beginPaint();

    _sculptedZones = std::unordered_set<Ref<ZoneResource>>();
}

void HoleTool::endPaint() {
    ToolBase::endPaint();

    _sculptedZones = std::unordered_set<Ref<ZoneResource>>();

    _terraBrush->clearObjects();
    _terraBrush->createObjects();
}

Ref<ImageTexture> HoleTool::getToolCurrentImageTexture(Ref<ZoneResource> zone) {
    return zone->get_heightMapTexture();
}

void HoleTool::paint(TerrainToolType toolType, Ref<Image> brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
    forEachBrushPixel(brushImage, brushSize, imagePosition, ([&](ImageZoneInfo &imageZoneInfo, float pixelBrushStrength) {
        Color currentPixel = imageZoneInfo.image->get_pixel(imageZoneInfo.zoneInfo.imagePosition.x, imageZoneInfo.zoneInfo.imagePosition.y);
        float holeValue = toolType == TerrainToolType::TERRAINTOOLTYPE_HOLEADD ? 1 : 0;

        Color newValue = Color(
            currentPixel.r,
            Math::max(0.0f, Math::min(1.0f, Math::lerp(currentPixel.g, holeValue, (float) Math::ceil(pixelBrushStrength) * (brushStrength * 100.0f)))),
            currentPixel.b,
            currentPixel.a
        );

        imageZoneInfo.image->set_pixel(imageZoneInfo.zoneInfo.imagePosition.x, imageZoneInfo.zoneInfo.imagePosition.y, newValue);
        _sculptedZones.insert(imageZoneInfo.zone);
    }));

    _terraBrush->get_terrainZones()->updateHeightmaps();
}